packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-debug-tracer.ts

```diff
@@ -133,13 +147,20 @@ export class VMDebugTracer {
   }
 
   private async _stepHandler(step: InterpreterStep, next: any) {
-    assertHardhatInvariant(
-      this._messages.length > 0,
-      "Step handler should be called after at least one beforeMessage handler"
-    );
+    try {
+      assertHardhatInvariant(
+        this._messages.length > 0,
+        "Step handler should be called after at least one beforeMessage handler"
+      );
 
-    const structLog = await this._stepToStructLog(step);
-    this._messages[this._messages.length - 1].structLogs.push(structLog);
+      const structLog = await this._stepToStructLog(step);
+      this._messages[this._messages.length - 1].structLogs.push(structLog);
+    } catch (e: any) {
+      // errors thrown in event handlers are lost, so we save this error to
+      // re-throw it in the `trace` function
```

Comment:
when is the `trace` function called? I guess this function is called from `trace`? Can you document this?

Reply:
Sorry, I don't get what you mean here. The `trace` function is the entry point of this class (the only public method), and that's where errors need to be re-thrown.